OpenRA.Mods.Common/Traits/World/HierarchicalPathFinder.cs

```diff
@@ -0,0 +1,836 @@
+				}
+
+				// If any nodes were left over they have no connections now, so we can remove them from the graph.
+				foreach (var unconnectedNode in abstractNodes)
+					abstractGraph.Remove(unconnectedNode);
+			}
+		}
+
+		/// <summary>
+		/// The abstract domains can become out of date when the abstract graph changes.
+		/// When this occurs, we must rebuild the domain cache.
+		/// </summary>
+		void RebuildDomains()
+		{
+			// First, rebuild the abstract graph if it is out of date.
+			RebuildDirtyGrids();
+
+			// Check if our domain cache is empty, if so this indicates it is out-of-date and needs rebuilding.
+			if (abstractDomains.Count != 0)
+				return;
+
+			List<GraphConnection> AbstractEdge(CPos abstractCell)
+			{
+				if (abstractGraph.TryGetValue(abstractCell, out var abstractEdge))
+					return abstractEdge;
+				return null;
+			}
+
+			// As in BuildGrid, flood fill the search graph until all disjoint domains are discovered.
+			var domain = 0u;
+			var abstractCells = abstractGraph.Keys.ToHashSet();
+			while (abstractCells.Count > 0)
+			{
+				var searchCell = abstractCells.First();
+				var search = PathSearch.ToTargetCellOverGraph(
+					AbstractEdge,
+					locomotor,
+					searchCell,
+					searchCell,
+					abstractGraph.Count / 8);
+				var searched = search.ExpandAll();
+				foreach (var abstractCell in searched)
+					abstractDomains.Add(abstractCell, domain);
+				abstractCells.ExceptWith(searched);
+				domain++;
+			}
+		}
+
+		/// <summary>
+		/// Performs a standard bi-directional A* search, but uses the abstract path to inform the heuristic.
+		/// </summary>
+		List<CPos> NavigatePathWithAbstractPathAsHeuristic(
+			Actor self, CPos source, CPos target, BlockedByActor check, Func<CPos, int> customCost, Actor ignoreActor, bool laneBias,
+			PathSearch abstractSearch, PathSearch reversedAbstractSearch)
+		{
+			Func<CPos, int> Heuristic(PathSearch search)
+			{
+				var graph = (SparsePathGraph)search.Graph;
+				return cell =>
+				{
+					// The search only explores accessible cells, so local cell is guaranteed reachable.
+					var gridInfo = gridInfos[GridIndex(cell)];
+					var abstractCell = gridInfo.AbstractCellForLocalCell(cell).Value;
+					var info = graph[abstractCell];
+
+					// Expand the abstract search if we have not visited the abstract cell.
+					if (info.Status == CellStatus.Unvisited)
+					{
+						search.TargetPredicate = c => c == abstractCell;
+						search.ExpandToTarget();
+						info = graph[abstractCell];
+
+						// The abstract cell is in fact unreachable - this indicates a bug.
+						if (info.Status == CellStatus.Unvisited)
+							throw new Exception("The abstract path should never be searched for an unreachable point.");
+					}
+
+					var cost = graph[info.PreviousNode].CostSoFar + costEstimator(cell, info.PreviousNode);
+					return cost;
+				};
+			}
+
+			// Note that we pair the search from source to target with the abstract search from target to search - the opposite.
```

Comment:
"the abstract search from target to search"? I assume this should be "from target to source"?